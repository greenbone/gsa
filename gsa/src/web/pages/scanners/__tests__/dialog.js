/* Copyright (C) 2019 Greenbone Networks GmbH
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */

import React from 'react';

import {setLocale} from 'gmp/locale/lang';
import Credential, {
  USERNAME_PASSWORD_CREDENTIAL_TYPE,
  CLIENT_CERTIFICATE_CREDENTIAL_TYPE,
} from 'gmp/models/credential';
import Scanner, {OSP_SCANNER_TYPE, GMP_SCANNER_TYPE} from 'gmp/models/scanner';

import {rendererWith, fireEvent} from 'web/utils/testing';

import ScannerDialog from 'web/pages/scanners/dialog';

setLocale('en');

const ospScanner = {
  _id: '1234',
  name: 'john',
  ca_pub: 'foo',
  credential: {
    _id: '5678',
  },
  comment: 'lorem ipsum',
  type: OSP_SCANNER_TYPE,
  host: 'mypc',
  port: '1357',
};

const gmpScanner = {
  _id: '1234',
  ca_pub: '',
  name: 'john',
  credential: {
    _id: '2345',
  },
  comment: 'lorem ipsum amet',
  type: GMP_SCANNER_TYPE,
  host: 'mypc',
};

const cred1 = Credential.fromElement({
  _id: '5678',
  name: 'foo',
  type: CLIENT_CERTIFICATE_CREDENTIAL_TYPE,
});

const cred2 = Credential.fromElement({
  _id: '2345',
  name: 'bar',
  type: USERNAME_PASSWORD_CREDENTIAL_TYPE,
});

// crendentials list is necessary along with 'type' to generate
// nonempty string items for the Select component for credential_id.
// Otherwise it would fail prop validation as '' is not a valid value
// for that component.
const credentials = [cred1, cred2];

const gmp = {settings: {enableGreenboneSensor: true}};

describe('ScannerDialog component tests', () => {
  test('should render', () => {
    const elem = {_id: 'foo', type: OSP_SCANNER_TYPE};
    const scanner = Scanner.fromElement(elem);
    const handleClose = jest.fn();
    const handleCredentialChange = jest.fn();
    const handleSave = jest.fn();
    const handleScannerTypeChange = jest.fn();

    const {render} = rendererWith({gmp});

    const {baseElement} = render(
      <ScannerDialog
        credential_id={'5678'}
        credentials={credentials}
        scanner={scanner}
        type={scanner.scannerType}
        onClose={handleClose}
        onCredentialChange={handleCredentialChange}
        onSave={handleSave}
        onScannerTypeChange={handleScannerTypeChange}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test('should display default info', () => {
    const scanner = Scanner.fromElement(ospScanner);

    const handleClose = jest.fn();
    const handleCredentialChange = jest.fn();
    const handleSave = jest.fn();
    const handleScannerTypeChange = jest.fn();

    const {render} = rendererWith({gmp});

    const {getAllByTestId, baseElement} = render(
      <ScannerDialog // using OSP Scanner to render the most amount of fields
        credentials={credentials}
        scanner={scanner}
        type={scanner.scannerType}
        credential_id={'5678'}
        onClose={handleClose}
        onCredentialChange={handleCredentialChange}
        onSave={handleSave}
        onScannerTypeChange={handleScannerTypeChange}
      />,
    );

    const inputs = baseElement.querySelectorAll('input');

    expect(inputs[0]).toHaveAttribute('name', 'name');
    expect(inputs[0]).toHaveAttribute('value', 'Unnamed'); // name field

    expect(inputs[1]).toHaveAttribute('name', 'comment');
    expect(inputs[1]).toHaveAttribute('value', ''); // comment field

    expect(inputs[2]).toHaveAttribute('name', 'host');
    expect(inputs[2]).toHaveAttribute('value', 'localhost');

    expect(inputs[3]).toHaveAttribute('name', 'port');
    expect(inputs[3]).toHaveAttribute('value', '9391');

    const selectedValues = getAllByTestId('select-selected-value');

    expect(selectedValues[0]).toHaveAttribute('title', 'OSP Scanner');
    expect(selectedValues[1]).toHaveAttribute('title', 'foo');
  });

  test('should display value from props', () => {
    const scanner = Scanner.fromElement(ospScanner);

    const handleClose = jest.fn();
    const handleCredentialChange = jest.fn();
    const handleSave = jest.fn();
    const handleScannerTypeChange = jest.fn();

    const {render} = rendererWith({gmp});

    const {getAllByTestId, baseElement} = render(
      <ScannerDialog // using OSP Scanner to render the most amount of fields
        comment={scanner.comment}
        credentials={credentials}
        credential_id={scanner.credential.id}
        host={scanner.host}
        name={scanner.name}
        port={scanner.port}
        scanner={scanner}
        type={scanner.scannerType}
        onClose={handleClose}
        onCredentialChange={handleCredentialChange}
        onSave={handleSave}
        onScannerTypeChange={handleScannerTypeChange}
      />,
    );

    const inputs = baseElement.querySelectorAll('input');

    expect(inputs[0]).toHaveAttribute('name', 'name');
    expect(inputs[0]).toHaveAttribute('value', 'john');

    expect(inputs[1]).toHaveAttribute('name', 'comment');
    expect(inputs[1]).toHaveAttribute('value', 'lorem ipsum');

    expect(inputs[2]).toHaveAttribute('name', 'host');
    expect(inputs[2]).toHaveAttribute('value', 'mypc');

    expect(inputs[3]).toHaveAttribute('name', 'port');
    expect(inputs[3]).toHaveAttribute('value', '1357');

    const selectedValues = getAllByTestId('select-selected-value');

    expect(selectedValues[0]).toHaveAttribute('title', 'OSP Scanner');
    expect(selectedValues[1]).toHaveAttribute('title', 'foo');
  });

  test('should save valid form state', () => {
    const scanner = Scanner.fromElement(gmpScanner);

    const handleClose = jest.fn();
    const handleCredentialChange = jest.fn();
    const handleSave = jest.fn();
    const handleScannerTypeChange = jest.fn();

    const {render} = rendererWith({gmp});

    const {getByTestId} = render(
      <ScannerDialog
        comment={scanner.comment}
        credential_id={scanner.credential.id}
        credentials={credentials}
        host={scanner.host}
        id={scanner.id}
        name={scanner.name}
        scanner={scanner}
        type={scanner.scannerType}
        onClose={handleClose}
        onCredentialChange={handleCredentialChange}
        onSave={handleSave}
        onScannerTypeChange={handleScannerTypeChange}
      />,
    );

    const saveButton = getByTestId('dialog-save-button');

    fireEvent.click(saveButton);

    expect(handleSave).toHaveBeenCalledWith({
      ca_pub: undefined,
      host: 'mypc',
      name: 'john',
      comment: 'lorem ipsum amet',
      credential_id: '2345',
      type: 4,
      id: '1234',
      port: '9391',
      which_cert: undefined,
    });
  });

  test('should change fields in create dialog', () => {
    const scanner = Scanner.fromElement(gmpScanner);

    const handleClose = jest.fn();
    const handleCredentialChange = jest.fn();
    const handleSave = jest.fn();
    const handleScannerTypeChange = jest.fn();

    const {render} = rendererWith({gmp});

    const {baseElement, getByName, getByTestId} = render(
      <ScannerDialog
        comment={scanner.comment}
        credential_id={scanner.credential.id}
        credentials={credentials}
        host={scanner.host}
        id={scanner.id}
        name={scanner.name}
        scanner={scanner}
        type={scanner.scannerType}
        onClose={handleClose}
        onCredentialChange={handleCredentialChange}
        onSave={handleSave}
        onScannerTypeChange={handleScannerTypeChange}
      />,
    );

    const inputs = baseElement.querySelectorAll('input');

    expect(inputs[0]).toHaveAttribute('value', 'john');
    expect(inputs[1]).toHaveAttribute('value', 'lorem ipsum amet');

    const nameInput = getByName('name');
    fireEvent.change(nameInput, {target: {value: 'ipsum'}});

    const commentInput = getByName('comment');
    fireEvent.change(commentInput, {target: {value: 'lorem'}});

    const saveButton = getByTestId('dialog-save-button');
    fireEvent.click(saveButton);

    expect(handleSave).toHaveBeenCalledWith({
      ca_pub: undefined,
      host: 'mypc',
      name: 'ipsum',
      comment: 'lorem',
      credential_id: '2345',
      type: 4,
      id: '1234',
      port: '9391',
      which_cert: undefined,
    });
  });

  test('should allow to close the dialog', () => {
    const elem = {_id: 'foo', type: OSP_SCANNER_TYPE};
    const scanner = Scanner.fromElement(elem);
    const handleClose = jest.fn();
    const handleCredentialChange = jest.fn();
    const handleSave = jest.fn();
    const handleScannerTypeChange = jest.fn();

    const {render} = rendererWith({gmp});

    const {getByTestId} = render(
      <ScannerDialog
        credentials={credentials}
        credential_id={'5678'}
        type={scanner.scannerType}
        scanner={scanner}
        onClose={handleClose}
        onCredentialChange={handleCredentialChange}
        onSave={handleSave}
        onScannerTypeChange={handleScannerTypeChange}
      />,
    );

    const closeButton = getByTestId('dialog-close-button');

    fireEvent.click(closeButton);

    expect(handleClose).toHaveBeenCalled();
    expect(handleSave).not.toHaveBeenCalled();
  });
});
