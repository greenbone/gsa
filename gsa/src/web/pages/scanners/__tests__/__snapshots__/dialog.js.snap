// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScannerDialog component tests should render 1`] = `
.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c19 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c23 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c25 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: -5px;
}

.c25 > * {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c25 > * {
  margin-left: 5px;
}

.c24 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c31 {
  height: 16px;
  width: 16px;
  line-height: 16px;
}

.c31 * {
  height: inherit;
  width: inherit;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 10% auto;
  border: 0;
  outline: 0;
  width: 800px;
  height: px;
}

.c0 {
  position: fixed;
  font-family: Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;
  font-size: 1.1em;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: 0;
  background: rgba(102,102,102,0.5);
  z-index: 600;
  -webkit-transition: opacity 1s ease-in;
  transition: opacity 1s ease-in;
  width: 100%;
  height: 100%;
}

.c37 {
  width: 0;
  height: 0;
  cursor: nwse-resize;
  border-right: 20px solid transparent;
  border-bottom: 20px solid transparent;
  border-top: 20px solid #fff;
}

.c36 {
  position: absolute;
  bottom: 3px;
  right: 3px;
  width: 20px;
  height: 20px;
  background: repeating-linear-gradient( -45deg, transparent, transparent 2px, #000 2px, #000 3px );
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: inherit;
  padding: 0;
  background: #fff;
  box-shadow: 5px 5px 10px #787878;
  border-radius: 3px;
  border: 1px solid #787878;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 1px solid #519032;
  font-weight: bold;
  font-size: 12px;
  font-family: Verdana,sans-serif;
  color: #519032;
  cursor: pointer;
  border-radius: 2px;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4:hover {
  color: #fff;
  background: #519032;
}

.c5 {
  height: 24px;
  width: 24px;
  line-height: 24px;
}

.c5 * {
  height: inherit;
  width: inherit;
}

.c33 {
  display: inline-block;
  padding: 0 15px;
  color: #393637;
  text-align: center;
  vertical-align: middle;
  font-size: 11px;
  font-weight: bold;
  line-height: 30px;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  background-color: #fff;
  border-radius: 2px;
  border: 1px solid #aaaaaa;
  cursor: pointer;
  overflow: visible;
  z-index: 1;
}

.c33:focus,
.c33:hover {
  border: 1px solid #393637;
}

.c33:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background: #66c430;
  font-weight: bold;
  color: #fff;
}

.c33[disabled] {
  cursor: not-allowed;
  opacity: 0.65;
  box-shadow: none;
}

.c33 img {
  height: 32px;
  width: 32px;
  margin-top: 5px 10px 5px -10px;
  vertical-align: middle;
}

.c33:link {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #393637;
}

.c34 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c35 {
  border: 1px solid #787878;
  color: #519032;
  background: #87d050;
}

.c35:hover {
  color: #fff;
  background: #519032;
}

.c32 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-width: 1px 0 0 0;
  border-style: solid;
  border-color: #c8d3d9;
  margin-top: 15px;
  padding: 10px 20px 10px 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px 5px 5px 10px;
  margin-bottom: 15px;
  border-radius: 2px 2px 0 0;
  border-bottom: 1px solid #787878;
  color: #fff;
  font-weight: bold;
  background: #66c430;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
}

.c7 {
  overflow: auto;
  padding: 0 15px;
  width: 100%;
  height: 100%;
  max-height: 400px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: hidden;
  height: 100%;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  padding-bottom: 10px;
}

.c10 {
  display: inline-block;
  max-width: 100%;
  font-weight: bold;
  text-align: right;
  padding-left: 10px;
  padding-right: 10px;
  width: 16.66666667%;
  margin-left: 0;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 83.33333333%;
  padding-left: 10px;
  padding-right: 10px;
}

.c22 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  width: 83.33333333%;
  padding-left: 10px;
  padding-right: 10px;
}

.c20 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  margin: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c21 {
  height: 16px;
  width: 16px;
  line-height: 16px;
}

.c21 * {
  height: inherit;
  width: inherit;
}

.c17 {
  border: 1px solid #aaaaaa;
  border-radius: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 1px 5px;
  background-color: #fff;
  color: #000;
  font-weight: normal;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 180px;
}

.c14 {
  color: #a94442;
  font-weight: bold;
  font-size: 19px;
  padding-bottom: 1px;
  padding-left: 4px;
  display: none;
}

.c18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  word-break: keep-all;
  white-space: nowrap;
  overflow: hidden;
  cursor: pointer;
  cursor: default;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c12 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  display: block;
  height: 22px;
  color: #393637;
  background-color: #fff;
  background-image: none;
  border: 1px solid #aaaaaa;
  border-radius: 2px;
  padding: 1px 8px;
}

.c12:-webkit-autofill {
  box-shadow: 0 0 0 1000px white inset;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c26 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: normal;
  cursor: pointer;
}

.c27 {
  font-family: inherit;
  font-size: inherit;
  padding: 0;
  margin: 0;
  margin-left: 10px;
  line-height: normal;
  width: auto;
  height: auto;
}

.c29 {
  opacity: 1;
}

.c28 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c30 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<body>
  <div
    id="portals"
  >
    <div>
      <div
        class="c0"
      >
        <div
          class="c1"
          height="px"
          role="dialog"
          tabindex="1"
          width="800px"
        >
          <div
            class="c2"
          >
            <div
              class="c3"
              data-testid="dialog-title-bar"
            >
              <span>
                New Scanner
              </span>
              <div
                class="c4 c5"
                data-testid="dialog-title-close-button"
                title="Close"
              >
                ×
              </div>
            </div>
            <div
              class="c6"
            >
              <div
                class="c7"
                data-testid="save-dialog-content"
              >
                <div
                  class="c8"
                >
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Name
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="10"
                    >
                      <input
                        class="c12 c13"
                        name="name"
                        type="text"
                        value="Unnamed"
                      />
                      <div
                        class="c14"
                        data-testid="error-marker"
                      >
                        ×
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Comment
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="10"
                    >
                      <input
                        class="c12 c13"
                        name="comment"
                        type="text"
                        value=""
                      />
                      <div
                        class="c14"
                        data-testid="error-marker"
                      >
                        ×
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Type
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="10"
                    >
                      <div
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-labelledby="downshift-0-label"
                        class="c15"
                        role="combobox"
                      >
                        <div
                          class="c16"
                          width="180px"
                        >
                          <div
                            aria-haspopup="true"
                            aria-label="open menu"
                            class="c17"
                            data-toggle="true"
                            role="button"
                            type="button"
                          >
                            <div
                              class="c18"
                              data-testid="select-selected-value"
                              title="OSP Scanner"
                            >
                              OSP Scanner
                            </div>
                            <div
                              class="c19"
                            >
                              <span
                                class="c20 c21"
                                data-testid="select-open-button"
                              >
                                ▼
                              </span>
                            </div>
                          </div>
                        </div>
                        <div
                          class="c14"
                          data-testid="error-marker"
                        >
                          ×
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Host
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="10"
                    >
                      <input
                        class="c12 c13"
                        name="host"
                        type="text"
                        value="localhost"
                      />
                      <div
                        class="c14"
                        data-testid="error-marker"
                      >
                        ×
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Port
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="10"
                    >
                      <input
                        class="c12 c13"
                        name="port"
                        type="text"
                        value="9391"
                      />
                      <div
                        class="c14"
                        data-testid="error-marker"
                      >
                        ×
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      CA Certificate
                    </label>
                    <div
                      class="c22"
                      data-testid="formgroup-content"
                      size="10"
                    >
                      <div
                        class="c23"
                      >
                        <div
                          class="c24"
                        >
                          <div
                            class="c25"
                            margin="5px"
                          >
                            <div
                              class="c23"
                            >
                              <label
                                class="c26"
                              >
                                <div
                                  class="c24"
                                >
                                  <div
                                    class="c25"
                                    margin="5px"
                                  >
                                    <input
                                      class="c27 c28"
                                      data-testid="radio-input"
                                      name="which_cert"
                                      type="radio"
                                      value="default"
                                    />
                                    <span
                                      class="c29"
                                      data-testid="radio-title"
                                    >
                                      Default
                                    </span>
                                  </div>
                                </div>
                              </label>
                              <label
                                class="c26"
                              >
                                <div
                                  class="c24"
                                >
                                  <div
                                    class="c25"
                                    margin="5px"
                                  >
                                    <input
                                      class="c27 c28"
                                      data-testid="radio-input"
                                      name="which_cert"
                                      type="radio"
                                      value="new"
                                    />
                                    <span
                                      class="c29"
                                      data-testid="radio-title"
                                    >
                                      New:
                                    </span>
                                  </div>
                                </div>
                              </label>
                            </div>
                            <input
                              class="c30"
                              disabled=""
                              name="ca_pub"
                              type="file"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Credential
                    </label>
                    <div
                      class="c22"
                      data-testid="formgroup-content"
                      size="10"
                    >
                      <div
                        class="c24"
                      >
                        <div
                          class="c25"
                          margin="5px"
                        >
                          <div
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            aria-labelledby="downshift-1-label"
                            class="c15"
                            role="combobox"
                          >
                            <div
                              class="c16"
                              width="180px"
                            >
                              <div
                                aria-haspopup="true"
                                aria-label="open menu"
                                class="c17"
                                data-toggle="true"
                                role="button"
                                type="button"
                              >
                                <div
                                  class="c18"
                                  data-testid="select-selected-value"
                                  title="foo"
                                >
                                  foo
                                </div>
                                <div
                                  class="c19"
                                >
                                  <span
                                    class="c20 c21"
                                    data-testid="select-open-button"
                                  >
                                    ▼
                                  </span>
                                </div>
                              </div>
                            </div>
                            <div
                              class="c14"
                              data-testid="error-marker"
                            >
                              ×
                            </div>
                          </div>
                          <div
                            class="c23"
                          >
                            <span
                              class="c31"
                              data-testid="svg-icon"
                              title="Create a new Credential"
                            >
                              <svg
                                title="Create a new Credential"
                              >
                                new.svg
                              </svg>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="c32"
            >
              <button
                class="c33 c34 c35"
                data-testid="dialog-close-button"
                title="Cancel"
              >
                Cancel
              </button>
              <button
                class="c33 c34 c35"
                data-testid="dialog-save-button"
                title="Save"
              >
                Save
              </button>
            </div>
          </div>
          <div
            class="c36"
          >
            <div
              class="c37"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;
